lib/src/optimism/mod.rs

```diff
@@ -236,129 +284,185 @@ impl<D: BatcherDb> DeriveMachine<D> {
+                    // A valid receipt should be provided for block building results
+                    let builder_journal =
+                        to_vec(&output).expect("Failed to encode builder journal");
+                    env::verify(
+                        Digest::from(self.derive_input.block_image_id),
+                        &builder_journal,
+                    )
+                    .expect("Failed to validate block build output");
+                    output
                 };
 
-                let new_op_head_hash = new_op_head.hash();
-
-                #[cfg(not(target_os = "zkvm"))]
-                info!(
-                    "Derived Op block {} w/ hash {}",
-                    new_op_head.number, new_op_head_hash
+                // Ensure that the output came from the expected input
+                ensure!(
+                    new_op_head_input.state_input.hash() == op_block_output.state_input_hash(),
+                    "Invalid state input hash"
                 );
+                match op_block_output {
+                    BlockBuildOutput::SUCCESS {
+                        hash: new_block_hash,
+                        head: new_block_head,
+                        ..
+                    } => {
+                        // obtain verified op block header
+                        #[cfg(not(target_os = "zkvm"))]
+                        log::info!(
+                            "Derived Op block {} w/ hash {}",
+                            new_block_head.number,
+                            new_block_hash
+                        );
 
-                self.op_batcher.state.safe_head = L2BlockInfo {
-                    hash: new_op_head_hash,
-                    timestamp: new_op_head.timestamp.try_into().unwrap(),
-                    l1_origin: BlockId {
-                        number: self.op_batcher.state.epoch.number,
-                        hash: self.op_batcher.state.epoch.hash,
-                    },
+                        self.op_batcher.state.safe_head = L2BlockInfo {
+                            hash: new_block_hash,
+                            timestamp: new_block_head.timestamp.try_into().unwrap(),
+                            l1_origin: BlockId {
+                                number: self.op_batcher.state.epoch.number,
+                                hash: self.op_batcher.state.epoch.hash,
+                            },
+                        };
+
+                        derived_op_blocks.push(BlockId {
+                            number: new_block_head.number,
+                            hash: new_block_hash,
+                        });
+                        self.op_head_block_header = new_block_head;
+
+                        if self.op_head_block_header.number == target_block_no {
+                            break;
+                        }
+                    }
+                    BlockBuildOutput::FAILURE { .. } => {
```

Comment:
Since it is not clear from the spec what should happen in this case. I think it would be just better to panic. Then, when this actually occurs in some time in practice we know that this needs to be revisited.

Reply:
I don't think it's worth modifying until we have more clarity then. For now, without panicking, at least we can avoid having invalid blocks lead to breaking the derivation process.

Reply:
OK, if you think so...
Let's create an issue for all this derivation ambiguity stuff, though, so we don't forget it.